### DISCUSSION TITLE
How to decompose the UFF (or MMFF94) scoring of a small molecule?

### DISCUSSION THREAD

**Discussion by UnixJunkie** [rdkit/rdkit#6377] [Q&A]
I have a ligand, I would like to score its current conformer using rdkit's UFF implementation.

Later, I would like to change some rotatable bond (single bond out of ring) and update
the conformer's energy bu just re-evaluating the part of the energy that is supposed to have
changed (i.e. the dihedral component).
Bond lengths, bond angles and partial charges being constant.

I suspect it should be faster than rescoring the conformer from scratch.

How to do this with rdkit?

Thanks a lot,
F.

**Comment by UnixJunkie** [comment]
@ptosco maybe?

**Reply by ptosco** [reply]
@UnixJunkie Currently UFF does not allow opting in/out individual energy terms; only MMFF does, e.g.
http://rdkit.org/docs/source/rdkit.ForceField.rdForceField.html?highlight=mmff#rdkit.ForceField.rdForceField.MMFFMolProperties.SetMMFFTorsionTerm

However, for your application I would rather compute the potential energy including all terms; the difference in terms of performance will be negligible, particularly as it is a single-point calculation.

**Comment by UnixJunkie** [comment]
Ok, good to know.
That might be one reason to switch to MMFF94.
Thanks a lot!

**Comment by ghutchis** [comment]
Single point energies on force fields are very fast.

As I mentioned on the e-mail list, this is unlikely to be much faster. You're forgetting that if you move a dihedral, you also have to re-calculate all the non-bonded terms. There are usually far more of those (approx. n^2) than bonds and angles. For MMFF94, you'd need to update both electrostatic and VdW terms. At that point, you're doing 90-95% of the single point evaluation anyway.

If you're evaluating lots of geometries and lots of molecules, the best approach is to generate multiple geometries and use multi-threading over the geometries and run molecules in batches / parallel. (Speaking as someone who's run ~3M molecules / 150M conformers recently.)

**Comment by UnixJunkie** [comment]
If this becomes necessary, I could decompose a ligand into rigid groups (they are separated by RotBonds).
Non-bonded interactions inside of a group can be computed once and for all.

**Reply by ghutchis** [reply]
It's up to you. Honestly, I'd consider that a bunch of programming time for not much speed benefit when you can multi-thread. But YMMV depending on use case.
